detect.go

```diff
@@ -0,0 +1,81 @@
+package composer
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/paketo-buildpacks/packit/v2"
+	"github.com/paketo-buildpacks/packit/v2/fs"
+	"github.com/paketo-buildpacks/packit/v2/scribe"
+)
+
+// PhpVersionResolverInterface provides a Resolve method to determine which Php Version to request
+//go:generate faux --interface PhpVersionResolverInterface --output fakes/php_version_resolver.go
+type PhpVersionResolverInterface interface {
+	Resolve(composerJsonPath, composerLockPath string) (version, versionSource string, err error)
+}
+
+func Detect(logEmitter scribe.Emitter, phpVersionResolver PhpVersionResolverInterface) packit.DetectFunc {
+	return func(context packit.DetectContext) (packit.DetectResult, error) {
+		composerJsonPath, composerLockPath, composerVar, composerVarFound := FindComposerFiles(context.WorkingDir)
+
+		if exists, err := fs.Exists(composerJsonPath); err != nil {
+			return packit.DetectResult{}, err
+		} else if !exists && !composerVarFound {
+			return packit.DetectResult{}, packit.Fail.WithMessage("no %s found", DefaultComposerJsonPath)
+		} else if !exists && composerVarFound {
+			return packit.DetectResult{}, packit.Fail.WithMessage("no %s found at location '%s'", DefaultComposerJsonPath, composerVar)
+		}
+
+		if exists, err := fs.Exists(composerLockPath); err != nil {
+			return packit.DetectResult{}, err
+		} else if !exists {
+			logEmitter.Title("WARNING: Include a 'composer.lock' file with your application! This will make sure the exact same version of dependencies are used when you build. It will also enable caching of your dependency layer.")
+		}
+
+		if composerVendorDir, found := os.LookupEnv(ComposerVendorDir); found {
+			if relativePath, err := filepath.Rel(context.WorkingDir, filepath.Join(context.WorkingDir, composerVendorDir)); err != nil {
+				return packit.DetectResult{}, err
+			} else if relativePath != composerVendorDir || strings.HasPrefix(relativePath, "..") {
+				return packit.DetectResult{}, packit.Fail.WithMessage("COMPOSER_VENDOR_DIR must be a relative path underneath the project root")
```

Comment:
Why is this a `packit.Fail` instead of an error?

Reply:
I used `packit.Fail.WithMessage` to indicate that this situation means that the buildpack cannot detect. It's possible the user has `COMPOSER_VENDOR_DIR` set to something perfectly valid for local development and tries to use that value with buildpacks and it won't work. 

Do you recommend something else such as `errors.New()`?

Reply:
I think of `packit.Fail` as a case where we want this buildpack to bow out of detection, but not cause an outright failure/error. Return an `error` says, let's break the build in the case where this happens. I think in a case where the app has a `composer.json` file but someone has set an environment variable in a breaking way should cause a failure if that makes sense. You'd know more about if that seems reasonable for the actual use case though

Reply:
Ok. I had expected that `return err` usually means something has happened outside of our control, and `return Packit.Fail` to mean "buildpack cannot participate". That was partially motivated by this wording:

```golang
// Fail is a sentinal value that can be used to indicate a failure to detect
// during the detect phase.
```

https://github.com/paketo-buildpacks/packit/blob/733d0983595cf912d4eb5521029e197bb489622a/fail.go#L5-L6

Reply:
Does that make sense? I will wait to hit `Resolve Conversation`.